src/metabase/models/collection.clj

```diff
@@ -1548,18 +1551,23 @@
                             child-type->parent-ids)))
                 (zipmap (keys child-type->parent-ids) (repeat #{}))
                 collections)]
-    (map (fn [{:keys [id] :as collection}]
-           (let [below (apply set/union
-                              (for [[child-type coll-id-set] child-type->ancestor-ids]
-                                (when (contains? coll-id-set id)
-                                  #{child-type})))
-                 here (into #{} (for [[child-type coll-id-set] child-type->parent-ids
-                                      :when (contains? coll-id-set id)]
-                                  child-type))]
-             (cond-> collection
-               (seq below) (assoc :below below)
-               (seq here) (assoc :here here))))
-         collections)))
+    (mapv (fn [{:keys [id] :as collection}]
+            (let [below (persistent!
+                         (reduce-kv (fn [acc child-type coll-id-set]
+                                      (cond-> acc
+                                        (contains? coll-id-set id) (conj! child-type)))
+                                    (transient #{})
+                                    child-type->ancestor-ids))
```

Comment:
Given that both `below` and `here` are calculated the same way given each mapping, it would be nice to extract a helper. Maybe call it `non-empty-types` or `present-types` or something.

Reply:
Done!